File: windows/src/Pomodoro.Core/UpdateChecker.cs
using System.Net.Http.Headers;
using System.Reflection;
using System.Runtime.InteropServices;
using System.Security.Cryptography;
using System.Text.Json.Nodes;

namespace Pomodoro.Core;

/// <summary>A release newer than the running build.</summary>
public sealed record AvailableUpdate
{
    public required string Version { get; init; }
    public required string Name { get; init; }
    public required string PageUrl { get; init; }

    /// <summary>The platform's own asset, when the release carries one.</summary>
    public string? DownloadUrl { get; init; }

    /// <summary>The published checksum, so a download can be verified.</summary>
    public string? ChecksumUrl { get; init; }
}

/// <summary>
/// Checks GitHub for a newer release on launch, and can install one.
///
/// Nothing happens without a click: the check only reports, and installing is
/// a deliberate action. What it downloads is always verified against the
/// published checksum before anything is run.
/// </summary>
public sealed class UpdateChecker
{
    public const string Repository = "dukunuu/pomodoro";

    private static readonly HttpClient Client = new() { Timeout = TimeSpan.FromSeconds(15) };

    /// <summary>Separate client: an installer download needs far longer than an API call.</summary>
    private static readonly HttpClient Downloader = new() { Timeout = TimeSpan.FromMinutes(15) };

    private static string SettingsPath => Path.Combine(DataPaths.Directory, "pomodoro-updates.json");

    public AvailableUpdate? Available { get; private set; }
    public bool Checking { get; private set; }
    public string? LastError { get; private set; }

    public event Action? Changed;

    /// <summary>
    /// The build's marketing version. A source build reports 0.0.0-dev, which
    /// compares older than any release — correct, if a little eager.
    /// </summary>
    public static string CurrentVersion
    {
        get
        {
            var info = Assembly.GetEntryAssembly()?
                .GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
            if (string.IsNullOrEmpty(info)) return "0.0.0";
            // The SDK appends "+<commit sha>" to the informational version.
            var plus = info.IndexOf('+');
            return plus > 0 ? info[..plus] : info;
        }
    }

    public bool Enabled
    {
        get => ReadRaw()?["enabled"]?.GetValue<bool>() ?? true;
        set => WriteSettings(value, LastCheck);
    }

    private DateTimeOffset? LastCheck
    {
        get
        {
            var raw = ReadRaw()?["lastCheck"]?.GetValue<string>();
            return DateTimeOffset.TryParse(raw, out var value) ? value : null;
        }
    }

    private static JsonObject? ReadRaw()
    {
        var text = AtomicFile.Read(SettingsPath);
        if (text is null) return null;
        try { return JsonNode.Parse(text) as JsonObject; }
        catch (System.Text.Json.JsonException) { return null; }
    }

    private static void WriteSettings(bool enabled, DateTimeOffset? lastCheck)
    {
        var payload = new JsonObject
        {
            ["version"] = 1,
            ["enabled"] = enabled,
            ["lastCheck"] = lastCheck?.ToString("o")
        };
        AtomicFile.Write(SettingsPath, Persistence.Json(payload));
    }

    /// <summary>Called at launch. Quiet: no dialog, no error unless asked.</summary>
    public async Task CheckOnLaunchAsync()
    {
        if (!Enabled) return;
        // Once a day is plenty for a pomodoro timer, and keeps the API's
        // unauthenticated rate limit far out of reach.
        var last = LastCheck;
        if (last is not null && DateTimeOffset.UtcNow - last.Value < TimeSpan.FromDays(1)) return;
        await CheckAsync(force: false).ConfigureAwait(false);
    }

    public async Task<AvailableUpdate?> CheckAsync(bool force)
    {
        if (Checking) return Available;
        Checking = true;
        LastError = null;
        Changed?.Invoke();

        try
        {
            using var request = new HttpRequestMessage(
                HttpMethod.Get, $"https://api.github.com/repos/{Repository}/releases/latest");
            request.Headers.Accept.Add(
                new MediaTypeWithQualityHeaderValue("application/vnd.github+json"));
            request.Headers.UserAgent.ParseAdd($"Pomodoro/{CurrentVersion}");

            using var response = await Client.SendAsync(request).ConfigureAwait(false);
            if (response.StatusCode == System.Net.HttpStatusCode.NotFound)
            {
                // No published release yet; not an error worth showing.
                WriteSettings(Enabled, DateTimeOffset.UtcNow);
                return null;
            }
            if (!response.IsSuccessStatusCode)
            {
                LastError = $"GitHub returned HTTP {(int)response.StatusCode}.";
                return null;
            }

            var body = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
            if (JsonNode.Parse(body) is not JsonObject json)
            {
                LastError = "GitHub returned unreadable JSON.";
                return null;
            }

            WriteSettings(Enabled, DateTimeOffset.UtcNow);
            if (json["draft"]?.GetValue<bool>() == true ||
                json["prerelease"]?.GetValue<bool>() == true)
            {
                return null;
            }

            var tag = json["tag_name"]?.GetValue<string>();
            var page = json["html_url"]?.GetValue<string>();
            if (string.IsNullOrEmpty(tag) || string.IsNullOrEmpty(page)) return null;

            var latest = Normalize(tag);
            if (!IsNewer(latest, CurrentVersion))
            {
                Available = null;
                return null;
            }

            // The installer for this architecture, plus its checksum: an
            // update that installs itself must verify what it downloaded.
            var suffix = RuntimeInformation.ProcessArchitecture == Architecture.Arm64
                ? "-windows-arm64.exe"
                : "-windows-x64.exe";
            string? asset = null;
            string? checksum = null;
            foreach (var item in (json["assets"] as JsonArray) ?? [])
            {
                var name = (item as JsonObject)?["name"]?.GetValue<string>();
                var url = (item as JsonObject)?["browser_download_url"]?.GetValue<string>();
                if (name is null || url is null) continue;
                if (name.EndsWith(suffix, StringComparison.OrdinalIgnoreCase)) asset = url;
                else if (name.EndsWith(suffix + ".sha256", StringComparison.OrdinalIgnoreCase)) checksum = url;
            }

            Available = new AvailableUpdate
            {
                Version = latest,
                Name = json["name"]?.GetValue<string>() is { Length: > 0 } named ? named : tag,
                PageUrl = page,
                DownloadUrl = asset,
                ChecksumUrl = checksum
            };
            return Available;
        }
        catch (Exception error) when (error is HttpRequestException or TaskCanceledException)
        {
            if (force) LastError = error.Message;
            return null;
        }
        finally
        {
            Checking = false;
            Changed?.Invoke();
        }
    }

    /// <summary>
    /// Downloads the installer to a temporary file and verifies it against the
    /// published checksum. Returns the path, or throws with a reason.
    /// A self-installing update must never run something it has not checked.
    /// </summary>
    public static async Task<string> DownloadAsync(
        AvailableUpdate update,
        IProgress<double>? progress = null,
        CancellationToken cancellation = default)
    {
        if (string.IsNullOrEmpty(update.DownloadUrl))
        {
            throw new InvalidOperationException("This release has no installer for your architecture.");
        }

        var directory = Path.Combine(Path.GetTempPath(), "Pomodoro-update");
        System.IO.Directory.CreateDirectory(directory);
        var target = Path.Combine(directory, $"Pomodoro-{update.Version}-setup.exe");

        using (var response = await Downloader.GetAsync(update.DownloadUrl,
                   HttpCompletionOption.ResponseHeadersRead, cancellation).ConfigureAwait(false))
        {
            response.EnsureSuccessStatusCode();
            var total = response.Content.Headers.ContentLength ?? 0;
            await using var source = await response.Content.ReadAsStreamAsync(cancellation)
                .ConfigureAwait(false);
            await using var file = File.Create(target);

            var buffer = new byte[81920];
            long read = 0;
            int count;
            while ((count = await source.ReadAsync(buffer, cancellation).ConfigureAwait(false)) > 0)
            {
                await file.WriteAsync(buffer.AsMemory(0, count), cancellation).ConfigureAwait(false);
                read += count;
                if (total > 0) progress?.Report((double)read / total);
            }
        }

        if (!string.IsNullOrEmpty(update.ChecksumUrl))
        {
            var published = (await Downloader.GetStringAsync(update.ChecksumUrl, cancellation)
                .ConfigureAwait(false)).Trim().Split(' ')[0].ToLowerInvariant();
            await using var stream = File.OpenRead(target);
            var actual = Convert.ToHexString(
                await SHA256.HashDataAsync(stream, cancellation).ConfigureAwait(false)).ToLowerInvariant();
            if (!string.Equals(published, actual, StringComparison.Ordinal))
            {
                File.Delete(target);
                throw new InvalidOperationException(
                    "The downloaded installer did not match its published checksum.");
            }
        }

        return target;
    }

    /// <summary>Strips a leading "v" and any pre-release suffix: v1.2.3-beta → 1.2.3.</summary>
    public static string Normalize(string? raw)
    {
        var text = (raw ?? string.Empty).Trim();
        if (text.StartsWith('v')) text = text[1..];
        var dash = text.IndexOf('-');
        return dash > 0 ? text[..dash] : text;
    }

    public static bool IsNewer(string candidate, string current)
    {
        var left = Parse(candidate);
        var right = Parse(current);
        for (var index = 0; index < Math.Max(left.Length, right.Length); index++)
        {
            var a = index < left.Length ? left[index] : 0;
            var b = index < right.Length ? right[index] : 0;
            if (a != b) return a > b;
        }
        return false;
    }

    private static int[] Parse(string version) =>
        Normalize(version).Split('.')
            .Select(part => int.TryParse(part, out var value) ? value : 0)
            .ToArray();
}

File: macos/Sources/Pomodoro/Core/UpdateChecker.swift
import Foundation
import Combine

/// A release newer than the running build.
struct AvailableUpdate: Equatable {
    var version: String
    var name: String
    var pageURL: URL
    /// The platform's own asset, when the release carries one.
    var downloadURL: URL?
    /// The published checksum, so a download can be verified before it runs.
    var checksumURL: URL?
    var publishedAt: Date?
}

/// Checks GitHub for a newer release on launch.
///
/// It only ever *tells* you: downloading and installing stay a deliberate
/// click, because this app writes files the other front ends read and a
/// silent swap under a running timer is not worth the convenience.
@MainActor
final class UpdateChecker: ObservableObject {
    static let repository = "dukunuu/pomodoro"

    @Published private(set) var available: AvailableUpdate?
    @Published private(set) var checking = false
    @Published private(set) var lastError: String?

    private let defaults = UserDefaults.standard
    private let session = URLSession(configuration: .ephemeral)

    /// The build's marketing version. A source build reports 0.0.0-dev, which
    /// compares older than any release — correct, if a little eager.
    var currentVersion: String {
        (Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String) ?? "0.0.0"
    }

    var enabled: Bool {
        get { defaults.object(forKey: "checkForUpdates") == nil ? true : defaults.bool(forKey: "checkForUpdates") }
        set { defaults.set(newValue, forKey: "checkForUpdates") }
    }

    private var lastCheck: Date? {
        get { defaults.object(forKey: "lastUpdateCheck") as? Date }
        set { defaults.set(newValue, forKey: "lastUpdateCheck") }
    }

    /// Called at launch. Quiet: no alert, no error surfaced unless asked.
    func checkOnLaunch() {
        guard enabled else { return }
        // Once a day is plenty for a pomodoro timer, and keeps the API's
        // unauthenticated rate limit far out of reach.
        if let last = lastCheck, Date().timeIntervalSince(last) < 86_400 { return }
        Task { await check(force: false) }
    }

    @discardableResult
    func check(force: Bool) async -> AvailableUpdate? {
        guard !checking else { return available }
        checking = true
        lastError = nil
        defer { checking = false }

        var request = URLRequest(url: URL(string:
            "https://api.github.com/repos/\(Self.repository)/releases/latest")!)
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        request.setValue("Pomodoro/\(currentVersion)", forHTTPHeaderField: "User-Agent")
        request.timeoutInterval = 15

        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                lastError = "No response from GitHub."
                return nil
            }
            if http.statusCode == 404 {
                // No published release yet; not an error worth showing.
                lastCheck = Date()
                return nil
            }
            guard http.statusCode == 200 else {
                lastError = "GitHub returned HTTP \(http.statusCode)."
                return nil
            }
            guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                lastError = "GitHub returned unreadable JSON."
                return nil
            }

            lastCheck = Date()
            guard (json["draft"] as? Bool) != true, (json["prerelease"] as? Bool) != true else {
                return nil
            }
            guard let tag = json["tag_name"] as? String,
                  let pageString = json["html_url"] as? String,
                  let page = URL(string: pageString) else { return nil }

            let latest = Self.normalize(tag)
            guard Self.isNewer(latest, than: Self.normalize(currentVersion)) else {
                available = nil
                return nil
            }

            // The disk image and its checksum: an update that installs itself
            // must verify what it downloaded.
            var asset: URL?
            var checksum: URL?
            for item in (json["assets"] as? [[String: Any]]) ?? [] {
                guard let name = item["name"] as? String,
                      let urlString = item["browser_download_url"] as? String,
                      let url = URL(string: urlString) else { continue }
                if name.hasSuffix(".dmg") { asset = url }
                if name.hasSuffix(".dmg.sha256") { checksum = url }
            }

            let update = AvailableUpdate(
                version: latest,
                name: (json["name"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? tag,
                pageURL: page,
                downloadURL: asset,
                checksumURL: checksum,
                publishedAt: (json["published_at"] as? String).flatMap {
                    ISO8601DateFormatter().date(from: $0)
                }
            )
            available = update
            return update
        } catch {
            lastError = force ? error.localizedDescription : nil
            return nil
        }
    }

    // MARK: - Version comparison

    /// Strips a leading "v" and any pre-release suffix: "v1.2.3-beta" → "1.2.3".
    static func normalize(_ raw: String) -> String {
        var text = raw.trimmingCharacters(in: .whitespaces)
        if text.hasPrefix("v") { text.removeFirst() }
        if let dash = text.firstIndex(of: "-") { text = String(text[text.startIndex..<dash]) }
        return text
    }

    static func components(_ version: String) -> [Int] {
        version.split(separator: ".").map { Int($0) ?? 0 }
    }

    static func isNewer(_ candidate: String, than current: String) -> Bool {
        let left = components(candidate)
        let right = components(current)
        for index in 0..<max(left.count, right.count) {
            let a = index < left.count ? left[index] : 0
            let b = index < right.count ? right[index] : 0
            if a != b { return a > b }
        }
        return false
    }
}
